File: src/php/RequestUtils.php
<?php

declare(strict_types=1);

namespace Anonympins\Fingerprint;

use Anonympins\Fingerprint\FingerprintBuilder;

/**
 * Fournit des fonctions utilitaires pour extraire des informations et calculer
 * des scores de suspicion à partir d'un objet RequestContext.
 */
class RequestUtils
{
    /**
     * Base de données de fingerprints TLS (JA3) connus.
     * @var array<string, string|string[]>
     */
    private const TLS_FINGERPRINT_DB = [
        // --- Chrome (Desktop) ---
        'e188a442b87f422c5a1e80b05399435b' => 'Chrome',
        'd8e35855049321c6042a4325c697858f' => 'Chrome',
        'a9f90958d44533748c139a5d1895b925' => 'Chrome',
        '3b5379916d2b3882253c42885956a350' => 'Chrome',
        // --- Firefox (Desktop) ---
        'b386946a5a586163c7c533636b45c355' => 'Firefox',
        '66236495a523c1785f8f3a105b248b11' => 'Firefox',
        // --- Safari & iOS (Shared TLS Stack) ---
        'b633f21d532d35967c8753c38536b4d3' => 'Safari',
        '4d7a28d5f55b359b69100a311013f03e' => ['Safari', 'Chrome', 'Firefox'],
        // --- Common Libraries & Bots ---
        '47344a349b75c4e82333475553b5f358' => 'Python',
        'b29587b8a143c42546133ad7704b3310' => 'Go',
        'd435b5223b2884c5a832b842637e245f' => 'Java',
        'c72366b9551263d990b7fa574225332c' => 'curl',
    ];

    /**
     * Construit une empreinte composite côté serveur.
     */
    public static function getCompositeDeviceHash(RequestContext $context): string
    {
        $srv = new FingerprintBuilder();

        $clientFp = $context->getHeader('x-device-fingerprint');
        if ($clientFp && str_contains($clientFp, 'cvs:')) {
            $srv->add("client_fp_hash", $clientFp);
        }

        // 1. User Agent
        $ua = $context->getHeader("user-agent");
        if ($ua) {
            $srv->add("ua", $ua);
        }

        // 2. Signaux de bas niveau (fournis par le proxy)
        if ($context->ja3) $srv->add("ja3", $context->ja3);
        if ($context->ja4) $srv->add("ja4", $context->ja4);
        if ($context->http2Fingerprint) $srv->add("h2", $context->http2Fingerprint);
        if ($context->tcpFingerprint) $srv->add("tcp", $context->tcpFingerprint);

        // 3. Signaux de haut niveau (en-têtes)
        $headersToCapture = [
            "ch_ua" => "sec-ch-ua",
            "ch_platform" => "sec-ch-ua-platform",
            "ch_mobile" => "sec-ch-ua-mobile",
            "accept_lang" => "accept-language",
            "accept_enc" => "accept-encoding",
            "accept" => "accept"
        ];

        foreach ($headersToCapture as $key => $headerName) {
            $headerValue = $context->getHeader($headerName);
            if ($headerValue) {
                $srv->add($key, $headerValue);
            }
        }

        // 4. Signaux de contexte
        if ($context->httpVersion) {
            $srv->add("http_ver", $context->httpVersion);
        }
        if (!empty($context->cookies)) {
            $cookieKeys = array_keys($context->cookies);
            sort($cookieKeys);
            $srv->add("cookie_keys", implode(',', $cookieKeys));
        }

        return (string)$srv;
    }

    /**
     * Calcule un score basé sur les anomalies des en-têtes HTTP.
     */
    public static function getHeaderAnomalies(RequestContext $context): array
    {
        $anomalyScore = 0;
        $ua = $context->getHeader("user-agent");

        if (!$ua || strlen($ua) < 10) {
            $anomalyScore += 60;
        }
        if (!$context->getHeader("accept-language")) {
            $anomalyScore += 25;
        }
        if ($context->httpVersion === "1.0") {
            $anomalyScore += 15;
        }

        return ['headerAnomalyScore' => min(100, $anomalyScore)];
    }

    /**
     * Calcule un score d'incohérence entre le fingerprint TLS (JA3) et le User-Agent.
     */
    public static function getTlsSpoofingScore(RequestContext $context): array
    {
        $ja3 = $context->ja3;
        $ua = $context->getHeader("user-agent") ?? '';

        if ($ja3 && (strlen($ua) < 10 || str_contains(strtolower($ua), 'python') || str_contains(strtolower($ua), 'curl'))) {
            return ['tlsSpoofingScore' => 50];
        }

        if ($ja3 && $ua) {
            $expectedBrowsers = self::TLS_FINGERPRINT_DB[$ja3] ?? null;

            if ($expectedBrowsers) {
                if (!is_array($expectedBrowsers)) {
                    $expectedBrowsers = [$expectedBrowsers];
                }

                $claimedBrowserInfo = self::parseUserAgent($ua);
                $claimedBrowser = $claimedBrowserInfo['browser'] ?? null;

                if ($claimedBrowser) {
                    $isMatch = false;
                    foreach ($expectedBrowsers as $expected) {
                        if (str_starts_with($claimedBrowser, $expected)) {
                            $isMatch = true;
                            break;
                        }
                    }
                    if (!$isMatch) {
                        return ['tlsSpoofingScore' => 80];
                    }
                }
            }
        }

        return ['tlsSpoofingScore' => 0];
    }

    /**
     * Analyse basique d'un User-Agent.
     * @return array{browser?: string, os?: string, device?: string}
     */
    public static function parseUserAgent(string $ua): array
    {
        $result = [];

        // Détection du navigateur
        if (preg_match('/Chrome\/(\d+)/', $ua) && !str_contains($ua, 'Edg')) {
            $result['browser'] = 'Chrome';
        } elseif (preg_match('/Firefox\/(\d+)/', $ua)) {
            $result['browser'] = 'Firefox';
        } elseif (preg_match('/Safari\/(\d+)/', $ua) && !str_contains($ua, 'Chrome')) {
            $result['browser'] = 'Safari';
        } elseif (preg_match('/Edg\/(\d+)/', $ua)) {
            $result['browser'] = 'Edge';
        }

        // Détection de l'OS
        if (str_contains($ua, 'Windows NT 10.0')) $result['os'] = 'Windows';
        elseif (str_contains($ua, 'Mac OS X')) $result['os'] = 'macOS';
        elseif (str_contains($ua, 'Linux') && !str_contains($ua, 'Android')) $result['os'] = 'Linux';
        elseif (str_contains($ua, 'Android')) $result['os'] = 'Android';
        elseif (str_contains($ua, 'iPhone') || str_contains($ua, 'iPad')) $result['os'] = 'iOS';

        // Détection du type d'appareil
        if (str_contains($ua, 'Mobile')) $result['device'] = 'mobile';
        elseif (str_contains($ua, 'Tablet')) $result['device'] = 'tablet';
        else $result['device'] = 'desktop';

        return $result;
    }

    /**
     * Vérifie si une IP est privée.
     */
    public static function isPrivateIp(string $ip): bool
    {
        return filter_var(
            $ip,
            FILTER_VALIDATE_IP,
            FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE
        ) === false;
    }

    /**
     * Calcule un score basé sur l'incohérence temporelle entre client et serveur.
     */
    public static function getTimeInconsistencyScore(RequestContext $context): array
    {
        $behaviorHeader = $context->getHeader('x-behavior-metrics');
        if (!$behaviorHeader) {
            return ['timeInconsistencyScore' => 0];
        }

        $metrics = json_decode($behaviorHeader, true);
        $clientTimestamp = $metrics['clientTimestamp'] ?? null;

        if ($clientTimestamp && $context->requestTimestamp) {
            $timeDelta = $context->requestTimestamp - $clientTimestamp;
            $replayThresholdMs = 5000; // 5 secondes

            if ($timeDelta > $replayThresholdMs) {
                $score = min(100, ($timeDelta / $replayThresholdMs - 1) * 50);
                return ['timeInconsistencyScore' => $score];
            }
        }

        return ['timeInconsistencyScore' => 0];
    }
}

File: src/python/request_utils.py
import re

class RequestContext:
    def __init__(self, headers, http_version, http2_fingerprint=None, quic_fingerprint=None):
        self.headers = {k.lower(): v for k, v in headers.items()} if headers else {}
        self.http_version = http_version or "1.1"
        self.http2_fingerprint = http2_fingerprint or self.headers.get('x-http2-fingerprint')
        self.quic_fingerprint = quic_fingerprint or self.headers.get('x-quic-fp')

    def get_header(self, name):
        return self.headers.get(name.lower())

def parse_user_agent(ua):
    result = {"browser": None, "os": None, "device": "desktop"}
    if not ua:
        return result
    
    # Browser detection
    if "Chrome" in ua and "Edg" not in ua:
        result["browser"] = "Chrome"
        match = re.search(r"Chrome/(\d+)", ua)
        if match:
            result["browser"] += f"/{match.group(1)}"
    elif "Firefox" in ua:
        result["browser"] = "Firefox"
        match = re.search(r"Firefox/(\d+)", ua)
        if match:
            result["browser"] += f"/{match.group(1)}"
    elif "Safari" in ua and "Chrome" not in ua:
        result["browser"] = "Safari"
        match = re.search(r"Version/(\d+)", ua)
        if match:
            result["browser"] += f"/{match.group(1)}"
    elif "Edg" in ua:
        result["browser"] = "Edge"
        match = re.search(r"Edg/(\d+)", ua)
        if match:
            result["browser"] += f"/{match.group(1)}"

    # OS detection
    if "Windows NT 10.0" in ua:
        result["os"] = "Windows"
    elif "Mac OS X" in ua:
        result["os"] = "macOS"
    elif "Android" in ua:
        result["os"] = "Android"
    elif "iPhone" in ua or "iPad" in ua:
        result["os"] = "iOS"
    elif "Linux" in ua:
        result["os"] = "Linux"

    if "Mobile" in ua:
        result["device"] = "mobile"
    elif "Tablet" in ua:
        result["device"] = "tablet"

    return result

def get_protocol_anomaly_score(context):
    http2_anomaly = 0.0
    quic_anomaly = 0.0

    ua = context.get_header("user-agent") or ""
    ua_parts = parse_user_agent(ua)
    browser = ua_parts["browser"]

    # HTTP/2 Anomaly logic
    h2_fp = context.http2_fingerprint or context.get_header("x-http2-fingerprint")
    if h2_fp and browser:
        parts = h2_fp.split("|")
        if len(parts) >= 4:
            try:
                conn_window = int(parts[1])
            except ValueError:
                conn_window = 0
            header_order = parts[3]
            is_chromium = browser.startswith("Chrome") or browser.startswith("Edge")
            is_firefox = browser.startswith("Firefox")
            is_safari = browser.startswith("Safari")

            if is_chromium:
                if header_order and header_order != "m,a,s,p":
                    http2_anomaly += 60.0
                if conn_window in (65535, 65536):
                    http2_anomaly += 40.0
            elif is_firefox:
                if header_order and header_order != "m,s,p,a":
                    http2_anomaly += 60.0
            elif is_safari:
                if header_order and header_order != "m,s,p,a":
                    http2_anomaly += 60.0

    # QUIC Anomaly logic
    quic_fp = context.quic_fingerprint or context.get_header("x-quic-fp")
    if quic_fp and browser:
        parts = quic_fp.split(";")
        if len(parts) >= 2:
            params = {}
            for p in parts[1].split(","):
                kv = p.split("=", 1)
                if len(kv) == 2:
                    params[kv[0]] = kv[1]
            priority_order = parts[2] if len(parts) > 2 else ""

            if browser.startswith("Chrome") or browser.startswith("Edge"):
                try:
                    max_data = int(params.get("1", "0"))
                    max_streams = int(params.get("4", "0"))
                except ValueError:
                    max_data, max_streams = 0, 0

                if max_data > 0 and max_data < 1048576:
                    quic_anomaly += 40.0
                if max_streams > 0 and max_streams != 100:
                    quic_anomaly += 30.0
                if priority_order and "u=" not in priority_order:
                    quic_anomaly += 30.0
            elif browser.startswith("Firefox"):
                try:
                    max_data = int(params.get("1", "0"))
                except ValueError:
                    max_data = 0
                if max_data > 0 and max_data > 5000000:
                    quic_anomaly += 40.0

    return {
        "protocolAnomalyScore": max(
            0.0,
            min(100.0, http2_anomaly),
            min(100.0, quic_anomaly)
        )
    }